File: apps/mobile/ios/Sources/State/PersistenceManager.swift
//
//  PersistenceManager.swift
//  TchatApp
//
//  Created by Claude on 21/09/2024.
//

import Foundation

/// Manages local data persistence for the Tchat app
public class PersistenceManager {

    // MARK: - Properties
    private let userDefaults = UserDefaults.standard
    private let keychain = KeychainManager()

    // MARK: - UserDefaults Keys
    private enum Keys {
        static let themePreferences = "themePreferences"
        static let isAuthenticated = "isAuthenticated"
        static let currentUser = "currentUser"
        static let chatState = "chatState"
        static let storeState = "storeState"
        static let socialState = "socialState"
        static let videoState = "videoState"
        static let lastSyncTimestamp = "lastSyncTimestamp"
    }

    // MARK: - Public Methods

    /// Save codable object to UserDefaults
    public func save<T: Codable>(_ object: T, forKey key: String) {
        do {
            let data = try JSONEncoder().encode(object)
            userDefaults.set(data, forKey: key)
        } catch {
            print("Failed to save \(key): \(error)")
        }
    }

    /// Load codable object from UserDefaults
    public func load<T: Codable>(forKey key: String, type: T.Type = T.self) -> T? {
        guard let data = userDefaults.data(forKey: key) else { return nil }

        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            print("Failed to load \(key): \(error)")
            return nil
        }
    }

    /// Save string to UserDefaults
    public func save(_ string: String, forKey key: String) {
        userDefaults.set(string, forKey: key)
    }

    /// Load string from UserDefaults
    public func loadString(forKey key: String) -> String? {
        return userDefaults.string(forKey: key)
    }

    /// Save boolean to UserDefaults
    public func save(_ bool: Bool, forKey key: String) {
        userDefaults.set(bool, forKey: key)
    }

    /// Load boolean from UserDefaults
    public func loadBool(forKey key: String) -> Bool {
        return userDefaults.bool(forKey: key)
    }

    /// Save integer to UserDefaults
    public func save(_ int: Int, forKey key: String) {
        userDefaults.set(int, forKey: key)
    }

    /// Load integer from UserDefaults
    public func loadInt(forKey key: String) -> Int {
        return userDefaults.integer(forKey: key)
    }

    /// Save date to UserDefaults
    public func save(_ date: Date, forKey key: String) {
        userDefaults.set(date, forKey: key)
    }

    /// Load date from UserDefaults
    public func loadDate(forKey key: String) -> Date? {
        return userDefaults.object(forKey: key) as? Date
    }

    /// Remove object for key
    public func remove(forKey key: String) {
        userDefaults.removeObject(forKey: key)
    }

    /// Clear all stored data
    public func clearAll() {
        let keys = [
            Keys.themePreferences,
            Keys.isAuthenticated,
            Keys.currentUser,
            Keys.chatState,
            Keys.storeState,
            Keys.socialState,
            Keys.videoState,
            Keys.lastSyncTimestamp
        ]

        keys.forEach { userDefaults.removeObject(forKey: $0) }
        keychain.clearAll()
    }

    // MARK: - Secure Storage (Keychain)

    /// Save sensitive data to keychain
    public func saveSecure(_ data: Data, forKey key: String) -> Bool {
        return keychain.save(data, forKey: key)
    }

    /// Load sensitive data from keychain
    public func loadSecure(forKey key: String) -> Data? {
        return keychain.load(forKey: key)
    }

    /// Save secure string to keychain
    public func saveSecureString(_ string: String, forKey key: String) -> Bool {
        guard let data = string.data(using: .utf8) else { return false }
        return keychain.save(data, forKey: key)
    }

    /// Load secure string from keychain
    public func loadSecureString(forKey key: String) -> String? {
        guard let data = keychain.load(forKey: key) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    /// Remove secure data from keychain
    public func removeSecure(forKey key: String) -> Bool {
        return keychain.remove(forKey: key)
    }
}

// MARK: - Keychain Manager

/// Simple keychain manager for secure storage
public class KeychainManager {

    private let service = "com.tchat.app"

    /// Save data to keychain
    func save(_ data: Data, forKey key: String) -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecValueData as String: data
        ]

        // Delete existing item
        SecItemDelete(query as CFDictionary)

        // Add new item
        let status = SecItemAdd(query as CFDictionary, nil)
        return status == errSecSuccess
    }

    /// Load data from keychain
    func load(forKey key: String) -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]

        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        guard status == errSecSuccess else { return nil }
        return result as? Data
    }

    /// Remove data from keychain
    func remove(forKey key: String) -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key
        ]

        let status = SecItemDelete(query as CFDictionary)
        return status == errSecSuccess || status == errSecItemNotFound
    }

    /// Clear all keychain data for this service
    func clearAll() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service
        ]

        SecItemDelete(query as CFDictionary)
    }
}

// MARK: - Convenience Extensions

extension PersistenceManager {

    /// Save theme preferences
    public func saveThemePreferences(_ preferences: ThemePreferences) {
        save(preferences, forKey: Keys.themePreferences)
    }

    /// Load theme preferences
    public func loadThemePreferences() -> ThemePreferences? {
        return load(forKey: Keys.themePreferences, type: ThemePreferences.self)
    }

    /// Save user authentication state
    public func saveAuthenticationState(_ isAuthenticated: Bool) {
        save(isAuthenticated, forKey: Keys.isAuthenticated)
    }

    /// Load user authentication state
    public func loadAuthenticationState() -> Bool {
        return loadBool(forKey: Keys.isAuthenticated)
    }

    /// Save current user
    public func saveCurrentUser(_ user: UserModel) {
        save(user, forKey: Keys.currentUser)
    }

    /// Load current user
    public func loadCurrentUser() -> UserModel? {
        return load(forKey: Keys.currentUser, type: UserModel.self)
    }

    /// Save chat state
    public func saveChatState(_ state: ChatState) {
        save(state, forKey: Keys.chatState)
    }

    /// Load chat state
    public func loadChatState() -> ChatState? {
        return load(forKey: Keys.chatState, type: ChatState.self)
    }

    /// Save store state
    public func saveStoreState(_ state: StoreState) {
        save(state, forKey: Keys.storeState)
    }

    /// Load store state
    public func loadStoreState() -> StoreState? {
        return load(forKey: Keys.storeState, type: StoreState.self)
    }

    /// Save social state
    public func saveSocialState(_ state: SocialState) {
        save(state, forKey: Keys.socialState)
    }

    /// Load social state
    public func loadSocialState() -> SocialState? {
        return load(forKey: Keys.socialState, type: SocialState.self)
    }

    /// Save video state
    public func saveVideoState(_ state: VideoState) {
        save(state, forKey: Keys.videoState)
    }

    /// Load video state
    public func loadVideoState() -> VideoState? {
        return load(forKey: Keys.videoState, type: VideoState.self)
    }

    /// Save last sync timestamp
    public func saveLastSyncTimestamp(_ timestamp: Date) {
        save(timestamp, forKey: Keys.lastSyncTimestamp)
    }

    /// Load last sync timestamp
    public func loadLastSyncTimestamp() -> Date? {
        return loadDate(forKey: Keys.lastSyncTimestamp)
    }
}

File: apps/mobile/android/app/src/main/java/com/tchat/state/PersistenceManager.kt
package com.tchat.state

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import com.tchat.models.Workspace
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Manages local data persistence for the Tchat app
 */
class PersistenceManager(private val context: Context? = null) {

    // MARK: - Properties
    private val sharedPreferences: SharedPreferences? by lazy {
        context?.getSharedPreferences("tchat_prefs", Context.MODE_PRIVATE)
    }

    private val encryptedPreferences: SharedPreferences? by lazy {
        context?.let { ctx ->
            val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
            EncryptedSharedPreferences.create(
                "tchat_secure_prefs",
                masterKeyAlias,
                ctx,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        }
    }

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    // MARK: - SharedPreferences Keys
    companion object {
        private const val KEY_THEME_PREFERENCES = "theme_preferences"
        private const val KEY_IS_AUTHENTICATED = "is_authenticated"
        private const val KEY_CURRENT_USER = "current_user"
        private const val KEY_CHAT_STATE = "chat_state"
        private const val KEY_STORE_STATE = "store_state"
        private const val KEY_SOCIAL_STATE = "social_state"
        private const val KEY_VIDEO_STATE = "video_state"
        private const val KEY_LAST_SYNC_TIMESTAMP = "last_sync_timestamp"
    }

    // MARK: - Generic Methods

    /**
     * Save object as JSON string
     */
    private inline fun <reified T> saveObject(obj: T, key: String, secure: Boolean = false) {
        val jsonString = json.encodeToString(obj)
        val prefs = if (secure) encryptedPreferences else sharedPreferences
        prefs?.edit()?.putString(key, jsonString)?.apply()
    }

    /**
     * Load object from JSON string
     */
    private inline fun <reified T> loadObject(key: String, secure: Boolean = false): T? {
        val prefs = if (secure) encryptedPreferences else sharedPreferences
        val jsonString = prefs?.getString(key, null) ?: return null
        return try {
            json.decodeFromString<T>(jsonString)
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Save string value
     */
    fun saveString(key: String, value: String, secure: Boolean = false) {
        val prefs = if (secure) encryptedPreferences else sharedPreferences
        prefs?.edit()?.putString(key, value)?.apply()
    }

    /**
     * Load string value
     */
    fun loadString(key: String, secure: Boolean = false): String? {
        val prefs = if (secure) encryptedPreferences else sharedPreferences
        return prefs?.getString(key, null)
    }

    /**
     * Save boolean value
     */
    fun saveBoolean(key: String, value: Boolean) {
        sharedPreferences?.edit()?.putBoolean(key, value)?.apply()
    }

    /**
     * Load boolean value
     */
    fun loadBoolean(key: String, defaultValue: Boolean = false): Boolean {
        return sharedPreferences?.getBoolean(key, defaultValue) ?: defaultValue
    }

    /**
     * Save integer value
     */
    fun saveInt(key: String, value: Int) {
        sharedPreferences?.edit()?.putInt(key, value)?.apply()
    }

    /**
     * Load integer value
     */
    fun loadInt(key: String, defaultValue: Int = 0): Int {
        return sharedPreferences?.getInt(key, defaultValue) ?: defaultValue
    }

    /**
     * Save long value
     */
    fun saveLong(key: String, value: Long) {
        sharedPreferences?.edit()?.putLong(key, value)?.apply()
    }

    /**
     * Load long value
     */
    fun loadLong(key: String, defaultValue: Long = 0L): Long {
        return sharedPreferences?.getLong(key, defaultValue) ?: defaultValue
    }

    /**
     * Remove value for key
     */
    fun remove(key: String, secure: Boolean = false) {
        val prefs = if (secure) encryptedPreferences else sharedPreferences
        prefs?.edit()?.remove(key)?.apply()
    }

    /**
     * Clear all stored data
     */
    fun clearAll() {
        sharedPreferences?.edit()?.clear()?.apply()
        encryptedPreferences?.edit()?.clear()?.apply()
    }

    // MARK: - Specific State Methods

    /**
     * Save theme preferences
     */
    fun saveThemePreferences(preferences: ThemePreferences) {
        saveObject(preferences, KEY_THEME_PREFERENCES)
    }

    /**
     * Load theme preferences
     */
    fun loadThemePreferences(): ThemePreferences? {
        return loadObject<ThemePreferences>(KEY_THEME_PREFERENCES)
    }

    /**
     * Save user authentication state
     */
    fun saveAuthenticationState(isAuthenticated: Boolean) {
        saveBoolean(KEY_IS_AUTHENTICATED, isAuthenticated)
    }

    /**
     * Load user authentication state
     */
    fun loadAuthenticationState(): Boolean {
        return loadBoolean(KEY_IS_AUTHENTICATED)
    }

    /**
     * Save current user (encrypted)
     */
    fun saveCurrentUser(user: UserModel) {
        saveObject(user, KEY_CURRENT_USER, secure = true)
    }

    /**
     * Load current user (encrypted)
     */
    fun loadCurrentUser(): UserModel? {
        return loadObject<UserModel>(KEY_CURRENT_USER, secure = true)
    }

    /**
     * Save chat state
     */
    fun saveChatState(state: ChatState) {
        saveObject(state, KEY_CHAT_STATE)
    }

    /**
     * Load chat state
     */
    fun loadChatState(): ChatState? {
        return loadObject<ChatState>(KEY_CHAT_STATE)
    }

    /**
     * Save store state
     */
    fun saveStoreState(state: StoreState) {
        saveObject(state, KEY_STORE_STATE)
    }

    /**
     * Load store state
     */
    fun loadStoreState(): StoreState? {
        return loadObject<StoreState>(KEY_STORE_STATE)
    }

    /**
     * Save social state
     */
    fun saveSocialState(state: SocialState) {
        saveObject(state, KEY_SOCIAL_STATE)
    }

    /**
     * Load social state
     */
    fun loadSocialState(): SocialState? {
        return loadObject<SocialState>(KEY_SOCIAL_STATE)
    }

    /**
     * Save video state
     */
    fun saveVideoState(state: VideoState) {
        saveObject(state, KEY_VIDEO_STATE)
    }

    /**
     * Load video state
     */
    fun loadVideoState(): VideoState? {
        return loadObject<VideoState>(KEY_VIDEO_STATE)
    }

    /**
     * Save last sync timestamp
     */
    fun saveLastSyncTimestamp(timestamp: Long) {
        saveLong(KEY_LAST_SYNC_TIMESTAMP, timestamp)
    }

    /**
     * Load last sync timestamp
     */
    fun loadLastSyncTimestamp(): Long? {
        val timestamp = loadLong(KEY_LAST_SYNC_TIMESTAMP, -1L)
        return if (timestamp == -1L) null else timestamp
    }

    // MARK: - Secure Storage Methods

    /**
     * Save authentication token (encrypted)
     */
    fun saveAuthToken(token: String) {
        saveString("auth_token", token, secure = true)
    }

    /**
     * Load authentication token (encrypted)
     */
    fun loadAuthToken(): String? {
        return loadString("auth_token", secure = true)
    }

    /**
     * Remove authentication token
     */
    fun removeAuthToken() {
        remove("auth_token", secure = true)
    }

    /**
     * Save refresh token (encrypted)
     */
    fun saveRefreshToken(token: String) {
        saveString("refresh_token", token, secure = true)
    }

    /**
     * Load refresh token (encrypted)
     */
    fun loadRefreshToken(): String? {
        return loadString("refresh_token", secure = true)
    }

    /**
     * Remove refresh token
     */
    fun removeRefreshToken() {
        remove("refresh_token", secure = true)
    }

    /**
     * Save user credentials (encrypted)
     */
    fun saveUserCredentials(email: String, password: String) {
        saveString("user_email", email, secure = true)
        saveString("user_password", password, secure = true)
    }

    /**
     * Load user credentials (encrypted)
     */
    fun loadUserCredentials(): Pair<String?, String?> {
        val email = loadString("user_email", secure = true)
        val password = loadString("user_password", secure = true)
        return Pair(email, password)
    }

    /**
     * Remove user credentials
     */
    fun removeUserCredentials() {
        remove("user_email", secure = true)
        remove("user_password", secure = true)
    }

    // MARK: - Utility Methods

    /**
     * Check if key exists
     */
    fun contains(key: String, secure: Boolean = false): Boolean {
        val prefs = if (secure) encryptedPreferences else sharedPreferences
        return prefs?.contains(key) ?: false
    }

    /**
     * Get all keys
     */
    fun getAllKeys(secure: Boolean = false): Set<String> {
        val prefs = if (secure) encryptedPreferences else sharedPreferences
        return prefs?.all?.keys ?: emptySet()
    }

    /**
     * Get storage size (approximate)
     */
    fun getStorageSize(): Int {
        return (sharedPreferences?.all?.size ?: 0) + (encryptedPreferences?.all?.size ?: 0)
    }

    // MARK: - Model-specific save methods

    fun saveSession(session: UserSession) {
        saveObject(session, "current_session", secure = true)
    }

    fun saveWorkspace(workspace: Workspace) {
        saveObject(workspace, "current_workspace", secure = false)
    }

    fun loadSession(): UserSession? {
        return loadObject<UserSession>("current_session", secure = true)
    }

    fun loadWorkspace(): Workspace? {
        return loadObject<Workspace>("current_workspace", secure = false)
    }

    fun clearSession() {
        remove("current_session", secure = true)
        remove("current_workspace", secure = false)
    }
}